### PR TITLE
GEODE-9004: Align results of queries with and without map indexes

### PR DESCRIPTION
**Pull request by albertogpz**
Queries in which map fields are involved
in the condition, sometimes do not return the same entries
if map indexes are used, compared to when map indexes
are not used.

Differences were found when the condition on the
map field was of type '!=' and also when the condition
on the map field was comparing the value with null.

Example1:
Running a query with the following condition:
positions['SUN'] = null

in a region with the following entries:
entry1.positions=null
entry2.positions={'a'=>'b'}
entry3.positions={'SUN'=>null}

- will return entry1 and entry3 if there are no
  indexes defined.
- will return no entries if the following index is defined:
  positions['SUN','c']
- will return entry3 if the following index is defined:
  positions[*]

Example2:
Running a query with the following condition:
positions['SUN'] != '3'

in a region with the following entries:
entry1.positions=null
entry2.positions={'a'=>'b'}
entry3.positions={'SUN'=>'4'}
entry4.positions={'SUN'=>'3'}
entry5.positions={'SUN'=>null}

- will return all entries except for entry4 if:
  there are no indexes defined.
- will return entry3 if the following index is defined:
  positions['SUN','c']
- will return entry3 and entry5 if the following index is defined:
  positions[*]

In order to have the same results for these
queries no matter if indexes are used,
the following changes have been made:

- When using compact map indexes or map indexes
an index entry will be created for every entry
added to the region even if the entry does not
contain the indexed map (the map is null) or
it does not contain any of the indexed keys.
The above will not apply to indexes of type
"allkeys" (e.g. positions[*]).

- As a consequence of the statement above
about indexes of type "allkeys", when the index
configured is of "allkeys" type and the query
is of type "!=" (e.g. positions['SUN'] != "3"
or the query is comparing against a null value
(e.g. positions['SUN'] = null) the index will
not be used.
This is a limitation of this solution
as allowing the index to be used in this case
would require bigger and more complex changes that
have not been considered in this solution.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/integrationTest/java/org/apache/geode/cache/query/internal/index/IndexStatisticsJUnitTest.java
Comment (mhansonp) [inline geode-core/src/integrationTest/java/org/apache/geode/cache/query/internal/index/IndexStatisticsJUnitTest.java:351]: Where were these values changed?

**Comment by mhansonp** [comment]
Looks like stress new test failed. Please review the problems in the test run and make necessary changes.

**Comment by nabarunnag** [comment]
Running the tests locally, I was able to reproduce the failure. This is the gradle command used:
`./gradlew compileTestJava compileIntegrationTestJava compileDistributedTestJava repeatIntegrationTest --tests CompactRangeIndexJUnitTest --tests IndexMaintenanceJUnitTest --tests IndexStatisticsJUnitTest --tests MapRangeIndexMaintenanceJUnitTest --tests PRIndexStatisticsJUnitTest -PtestCount=5 -Prepeat=50 -PfailOnNoMatchingTests=false`

File: geode-core/src/integrationTest/java/org/apache/geode/cache/query/internal/index/IndexStatisticsJUnitTest.java
Comment (albertogpz) [reply geode-core/src/integrationTest/java/org/apache/geode/cache/query/internal/index/IndexStatisticsJUnitTest.java:351]: The reason for numbers doubling in size is that, in order to support the queries with "!=" and with null comparisons, the index needs to store the entry for each key in the index no matter if the entry contains the indexed keys in the map field.

For example, if the index is positions['DELL', 'YAHOO'] and an entry is added to the region for which positions is {'DELL' -> '3'], the index will contain two values:
- for 'DELL', for key='3' and the entry as value.
- for 'YAHOO' for key=IndexManager.NULL and the entry as value.

Prior to this change, the index only stored the first value and, as a consequence, queries like the following: positions['YAHOO'] = null did not return all the right results.

For those cases where 2 has been added, it is due to keys equal to IndexManager.NULL or Undefined. This has been corrected in the last commit so that these counters are aligned with the non compact indexes that do not count Null or undefined as keys.

**Comment by albertogpz** [comment]
> Running the tests locally, I was able to reproduce the failure. This is the gradle command used:
> `./gradlew compileTestJava compileIntegrationTestJava compileDistributedTestJava repeatIntegrationTest --tests CompactRangeIndexJUnitTest --tests IndexMaintenanceJUnitTest --tests IndexStatisticsJUnitTest --tests MapRangeIndexMaintenanceJUnitTest --tests PRIndexStatisticsJUnitTest -PtestCount=5 -Prepeat=50 -PfailOnNoMatchingTests=false`

Thanks for the command. I did manage to reproduce the tests with it.
I have pushed a new commit and now there are no failures.

**Review by mhansonp — APPROVED** [review]

File: geode-core/src/main/java/org/apache/geode/cache/query/internal/CompiledComparison.java
Comment (DonalEvans) [inline geode-core/src/main/java/org/apache/geode/cache/query/internal/CompiledComparison.java:660]: Would it be possible to extract the contents of this if condition to a method, and refactor it slightly to make it easier to parse and understand? As it currently is, it's very difficult to see what the desired condition is. Readability is more important than brevity here, I think.

**Review by DonalEvans — CHANGES_REQUESTED** [review]
A few requests for clarification of comments or code, and some small clean-up suggestions.

File: geode-core/src/main/java/org/apache/geode/cache/query/internal/CompiledComparison.java
Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/cache/query/internal/CompiledComparison.java:660]: Yes. I totally agree.

File: geode-core/src/integrationTest/java/org/apache/geode/cache/query/internal/index/MapRangeIndexMaintenanceJUnitTest.java
Comment (albertogpz) [reply geode-core/src/integrationTest/java/org/apache/geode/cache/query/internal/index/MapRangeIndexMaintenanceJUnitTest.java:610]: Cool!

File: geode-core/src/integrationTest/java/org/apache/geode/cache/query/internal/index/MapRangeIndexMaintenanceJUnitTest.java
Comment (albertogpz) [reply geode-core/src/integrationTest/java/org/apache/geode/cache/query/internal/index/MapRangeIndexMaintenanceJUnitTest.java:380]: Sure

File: geode-core/src/integrationTest/java/org/apache/geode/cache/query/internal/index/MapRangeIndexMaintenanceJUnitTest.java
Comment (albertogpz) [reply geode-core/src/integrationTest/java/org/apache/geode/cache/query/internal/index/MapRangeIndexMaintenanceJUnitTest.java:381]: Done

**Comment by onichols-pivotal** [comment]
looks like this is ready to merge? if more changes are coming, please convert to draft until then

**Comment by albertogpz** [comment]
> looks like this is ready to merge? if more changes are coming, please convert to draft until then

Only some minor documentation change was missing. Once it is approved, the PR can be merged.

File: geode-docs/developing/query_index/creating_map_indexes.html.md.erb
Comment (davebarnes97) [inline geode-docs/developing/query_index/creating_map_indexes.html.md.erb:63]: Please insert a comma after "index":  ...making use of the index, which could...

**Review by davebarnes97 — APPROVED** [review]
Doc review: Looks good to me. One punctuation change requested -- see comment.

**Pull request merged**
